### ISSUE TITLE
`AVG`, `VAR_*`, and `STDDEV*` functions promote integer and decimal types to floating-point, not numeric

### ISSUE THREAD

**Issue by vmarcos** [MaterializeInc/materialize#19577]
### What version of Materialize are you using?

v0.55.0 (448db5f3c)

### What is the issue?

When `AVG`, `VAR_*`, and `STDDEV*` functions are applied over integer and decimal types, Materialize promotes the output type to a floating-point representation, while PostgreSQL promotes to `numeric`. Our promotion choice leads to issues, since for example the output of constant folding and dataflow execution then diverge, as documented in #19511. 

To observe the issue, consider the following SQL statements, adapted from scenario 6) in #19511 :

```sql
CREATE TABLE t_dfr_vert (row_index INT, decimal_39_8_val DECIMAL(39,8));
INSERT INTO t_dfr_vert VALUES (6, 1.00000001::DECIMAL(39,8));
INSERT INTO t_dfr_vert VALUES (8, 0.99999999::DECIMAL(39,8));

CREATE OR REPLACE VIEW v_ctf_vert (row_index, decimal_39_8_val)
 AS SELECT 6, 1.00000001::DECIMAL(39,8)
    UNION SELECT 8, 0.99999999::DECIMAL(39,8);

CREATE VIEW dataflow_execution AS
SELECT
  stddev_pop(decimal_39_8_val)
FROM
  t_dfr_vert
WHERE row_index IN (8, 6);

CREATE VIEW constant_folding AS
SELECT
  stddev_pop(decimal_39_8_val)
FROM
  v_ctf_vert
WHERE row_index IN (8, 6);

SELECT * FROM dataflow_execution;
SELECT * FROM constant_folding;
```

The output in Materialize for the script above is:

```sql
CREATE TABLE
INSERT 0 1
INSERT 0 1
CREATE VIEW
CREATE VIEW
CREATE VIEW
       stddev_pop       
------------------------
 0.00017263348892818244
(1 row)

 stddev_pop 
------------
          0
(1 row)
```

In PostgreSQL, we get:

```sql
CREATE TABLE
INSERT 0 1
INSERT 0 1
CREATE VIEW
CREATE VIEW
CREATE VIEW
               stddev_pop               
----------------------------------------
 0.000000010000000000000000000000000000
(1 row)

               stddev_pop               
----------------------------------------
 0.000000010000000000000000000000000000
(1 row)
```

Note that the output types of the views differ across the two systems. In Materialize:

```sql
materialize=> \d dataflow_execution
                View "public.dataflow_execution"
   Column   |       Type       | Collation | Nullable | Default 
------------+------------------+-----------+----------+---------
 stddev_pop | double precision |           |          | 

materialize=> \d constant_folding
                 View "public.constant_folding"
   Column   |       Type       | Collation | Nullable | Default 
------------+------------------+-----------+----------+---------
 stddev_pop | double precision |           |          | 

```

However, in PostgreSQL:

```sql
vmarcos=# \d dataflow_execution
           View "public.dataflow_execution"
   Column   |  Type   | Collation | Nullable | Default 
------------+---------+-----------+----------+---------
 stddev_pop | numeric |           |          | 

vmarcos=# \d constant_folding
            View "public.constant_folding"
   Column   |  Type   | Collation | Nullable | Default 
------------+---------+-----------+----------+---------
 stddev_pop | numeric |           |          | 

```

This issue is a request to type promote to `numeric` in Materialize when we have `AVG`, `VAR_*`, and `STDDEV*` aggregations over integer and decimal data, so as to reduce our differences wrt. PostgreSQL and also fix the correctness problem of different results being produced under constant folding or dataflow execution in this case.

**Comment by vmarcos** [comment]
Fixing this issue is tricky due to the need to migrate data from pre-existing materialized views using `AVG` and `STDDEV*`. A general strategy to do this safely is discussed in this [Slack thread](https://materializeinc.slack.com/archives/C02FWJ94HME/p1685449031540389).

**Comment by uce** [comment]
Relevant discussion in [Slack](https://materializeinc.slack.com/archives/C02FWJ94HME/p1685449031540389).

**Comment by ParkMyCar** [comment]
@uce we chatted about this issue today in the Surfaces Team Meeting. Just wanted to double check, is this something you still need us to do, or is another team handling it?

**Comment by vmarcos** [comment]
hi there, @uce can confirm, but in our 1:1 this week, my understanding was that this issue would be handled by Surfaces. I had a draft PR, without the necessary migration, illustrating the path to a fix in #19527. Please feel free to push it forward, cherry-pick any of the commits, and/or start something new, as you find the most productive.

**Comment by nrainer-materialize** [comment]
As identified with https://github.com/MaterializeInc/materialize/pull/22967 (https://github.com/MaterializeInc/materialize/pull/22967/commits/54b3f84145aed540399d055934849895a87653b7 in particular), there are still inconsistency issues. After some digging through issues and PRs, my understanding is that they should be attributed to https://github.com/MaterializeInc/materialize/issues/15186. I will update the issue references in the output-consistency ignore definitions accordingly.
